### PR TITLE
[3152][ADD] account_invoice_delivered_amount

### PR DESCRIPTION
**Pull request by AungKoKoLin1997**
[3152](https://www.quartile.co/web#id=3152&action=771&model=project.task&view_type=form&menu_id=505)

### CODE REVIEW

**Comment by AungKoKoLin1997** [comment]
@kanda999  Can you please review first this PR?

File: account_invoice_delivered_amount/models/account_invoice_line.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+from odoo import fields, models, api
+
+class AccountInvoiceLine(models.Model):
+    _inherit = "account.invoice.line"
+
+    is_delivered = fields.Boolean(string="Deliverd", compute="_compute_delivered_amount", store=True)
+    deliverd_amount = fields.Monetary(string='Delivered Amount',
+        store=True, compute='_compute_delivered_amount')
+    
+    @api.depends("sale_line_ids", "product_id", "quantity")
```
Comment (yostashiro) [inline account_invoice_delivered_amount/models/account_invoice_line.py]: Listen to the specific field of sale.order.line.  We need to watch the delivered qty field.

**Review by yostashiro — CHANGES_REQUESTED** [review]
@AungKoKoLin1997 Haven't checked in detail, but we need a pre_init_hook for sure.  It is a large database.  Can you try and include the SQL to update the existing records?  If it's too complecated, please do not use much time on this and just create new columns in pre_init_hook to avoid computation at install.

**Review by kanda999 — CHANGES_REQUESTED** [review]
Function was not complete.
As tashiro-san commented, if the invoice was created and then delivered, the delivered totals and is_delivered needed to be updated.

**Review by kanda999 — CHANGES_REQUESTED** [review]
@AungKoKoLin1997 
I tried to install, but the existing data is not updated.  Am I making any mistakes?

**Comment by AungKoKoLin1997** [comment]
> @AungKoKoLin1997 I tried to install, but the existing data is not updated. Am I making any mistakes?

No. In the first installation, I made to avoid computing for large database.

> If it's too complecated, please do not use much time on this and just create new columns in pre_init_hook to avoid computation at install.

**Comment by kanda999** [comment]
> No. In the first installation, I made to avoid computing for large database.

Sorry I got it.  So how should existing data be updated?

**Comment by AungKoKoLin1997** [comment]
> So how should existing data be updated?

I still have no idea. I haven't never written sql for updating existing records at install and I am worried it will take much time. It will be nice if you know how to do it.

**Comment by kanda999** [comment]
@AungKoKoLin1997 
I asked the client about the need for existing data calicuration.

**Comment by AungKoKoLin1997** [comment]
> @AungKoKoLin1997 I asked the client about the need for existing data calicuration.

Sorry. I am not clear. Do you mean client also want to update data for the existing records?

**Comment by kanda999** [comment]
I mean I'm communicating if the client needs it or not.

File: account_invoice_delivered_amount/models/account_invoice_line.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+                    qty_delivered = sum([x.qty_delivered for x in line.sale_line_ids])
+                    if qty_delivered > 0:
+                        line.is_delivered = True
+                        line.deliverd_amount = qty_delivered * line.price_unit
+                        break
+                    line.is_delivered = False
+                    line.deliverd_amount = 0
+            else:
+                line.is_delivered = False
+                line.deliverd_amount = 0
```
Comment (yostashiro) [inline account_invoice_delivered_amount/models/account_invoice_line.py]:
```suggestion
    @api.depends("quantity", "price_unit", "sale_line_ids.qty_delivered")
    def _compute_delivered_amount(self):
        for line in self:
            if line.product_id.type == "service" or not line.sale_line_ids:
                line.is_delivered = True
                line.deliverd_amount = line.quantity * line.price_unit
                continue
            qty_delivered = sum(line.sale_line_ids.mapped("qty_delivered"))
            if qty_delivered > 0:
                line.is_delivered = True
                line.deliverd_amount = qty_delivered * line.price_unit
```
We will always treat services and credit notes as "delivered."

File: account_invoice_delivered_amount/models/account_invoice.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+    total_delivered_amount = fields.Monetary(
+        string="Total Delivered Amount",
+        currency_field="currency_id",
+        store=True,
+        compute="_compute_delivered_amount",
+    )
+    total_delivered_amount_signed = fields.Monetary(
+        string="Total Delivered Amount in Company Currency",
+        currency_field="company_currency_id",
```
Comment (yostashiro) [inline account_invoice_delivered_amount/models/account_invoice.py]:
```suggestion
```
Why `company_currency_id here`?

**Review by kanda999 — APPROVED** [review]
Function review

File: account_invoice_delivered_amount/views/account_invoice_views.xml [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+        </field>
+    </record>
+    <record id="view_account_invoice_tree" model="ir.ui.view">
+        <field name="name">account.invoice.tree</field>
+        <field name="model">account.invoice</field>
+        <field name="inherit_id" ref="account.invoice_tree" />
+        <field name="arch" type="xml">
+            <xpath expr="//field[@name='amount_total_signed']" position="before">
+                <field name="type" invisible="1" />
+                <field name="total_delivered_amount_signed" attrs="{'invisible': [('type','!=','out_invoice')]}" string="Total Delivered" sum="Total" />
```
Comment (yostashiro) [inline account_invoice_delivered_amount/views/account_invoice_views.xml]:
```suggestion
                <field name="total_delivered_amount_signed" string="Total Delivered" sum="Total" />
```
The field should show for credit notes as well.

File: account_invoice_delivered_amount/models/account_invoice.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+from odoo import api, fields, models
+
+
+class AccountInvoice(models.Model):
+    _inherit = "account.invoice"
+
+    total_delivered_amount = fields.Monetary(
+        string="Total Delivered Amount",
+        currency_field="currency_id",
```
Comment (yostashiro) [inline account_invoice_delivered_amount/models/account_invoice.py]:
```suggestion
```

File: account_invoice_delivered_amount/models/account_invoice.py [hunk 6]
```diff
@@ -0,0 +1,41 @@
+# Copyright 2023 Quartile Limited
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+from odoo import api, fields, models
+
+
+class AccountInvoice(models.Model):
+    _inherit = "account.invoice"
+
+    total_delivered_amount = fields.Monetary(
```
Comment (yostashiro) [inline account_invoice_delivered_amount/models/account_invoice.py]:
```suggestion
    amount_total_delivered = fields.Monetary(
```
Following the convention of other amount_xxx fields.

File: account_invoice_delivered_amount/models/account_invoice.py [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+        string="Total Delivered Amount",
+        currency_field="currency_id",
+        store=True,
+        compute="_compute_delivered_amount",
+    )
+    total_delivered_amount_signed = fields.Monetary(
```
Comment (yostashiro) [inline account_invoice_delivered_amount/models/account_invoice.py]:
```suggestion
    amount_total_delivered_signed = fields.Monetary(
```

File: account_invoice_delivered_amount/models/account_invoice.py [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+                and invoice.company_id
+                and invoice.currency_id != invoice.company_id.currency_id
+            ):
+                currency_id = invoice.currency_id.with_context(
+                    date=invoice.date_invoice
+                )
+                total_delivered_amount_signed = currency_id.compute(
+                    invoice.total_delivered_amount, invoice.company_id.currency_id
+                )
+            invoice.total_delivered_amount_signed = total_delivered_amount_signed
```
Comment (yostashiro) [inline account_invoice_delivered_amount/models/account_invoice.py]:
```suggestion
    @api.depends("invoice_line_ids.is_delivered", "invoice_line_ids.deliverd_amount", "type")
    def _compute_amount_delivered(self):
        for invoice in self:
            if invoice.type not in ["out_invoice", "out_refund"]:
                continue
            sign = -1 if self.type == "out_refund" else 1
            invoice.total_amount_delivered = sum(invoice.invoice_line_ids.mapped("deliverd_amount"))
            invoice.total_amount_delivered_signed = invoice.total_amount_delivered * sign
```

**Review by yostashiro — CHANGES_REQUESTED** [review]
Please extend the account invoice search for the tree view to allow filtering records by "Delivered".

File: account_invoice_delivered_amount/readme/DESCRIPTION.rst [hunk 9]
```diff
@@ -0,0 +1 @@
+This module displays delivered amount in customer invoices.
```
Comment (yostashiro) [inline account_invoice_delivered_amount/readme/DESCRIPTION.rst]:
```suggestion
This module adds delivered amount fields the customer invoice.

We regard service lines and credit notes as "delivered."

Background:

There are cases where user issues an invoice to the customer upfront in a business transaction,
just for the sake of convenience, even when the goods are yet to be delivered to them.
In such case, technically speaking, revenue should not be recognized until the delivery is done.

This module intends to make the "real" sales amount visible in the invoice views by computing
the delivered amount.
```

File: account_invoice_delivered_amount/models/account_invoice.py (hunk 3 above)
Comment (AungKoKoLin1997) [reply account_invoice_delivered_amount/models/account_invoice.py]: I think we should display the amount in company currency in tree view like other fields(eg . `amount_total_signed`). My intention for creating this field is to display with converted amount in company currency in tree view. If not, we don't even need this field. We can use only `total_delivered_amount` field.

Comment (yostashiro) [reply account_invoice_delivered_amount/models/account_invoice.py]: If I'm not mistaken, `amount_total_signed` (which shows in the list view) is not in company currency. `amount_total_company_signed` is.

Comment (AungKoKoLin1997) [reply account_invoice_delivered_amount/models/account_invoice.py]: Yes. I am confused with version 15. In v15, `amount_total_signed` is defined with company currency.https://github.com/odoo/odoo/blob/d4e322d4bad40f09fd11fd368c1f7e83e3bb7a36/addons/account/models/account_move.py#L246.
My intension for amount_total_signed was to display in company currency. So, should I create new field `amount_total_delivered_company_signed` for displaying with company currency in list view?

Comment (yostashiro) [reply account_invoice_delivered_amount/models/account_invoice.py]: > So, should I create new field amount_total_delivered_company_signed for displaying with company currency in list view?

No need to do that this time (there is no user demand for it).

**Review by yostashiro — APPROVED** [review]

File: account_invoice_delivered_amount/models/account_invoice_line.py [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+    delivered_amount = fields.Monetary(
+        string="Delivered Amount", store=True, compute="_compute_delivered_amount"
+    )
+
+    @api.depends("quantity", "price_unit", "sale_line_ids.qty_delivered")
+    def _compute_delivered_amount(self):
+        for line in self:
+            if line.product_id.type == "service" or not line.sale_line_ids:
+                line.is_delivered = True
+                line.delivered_amount = line.quantity * line.price_total
```
Comment (yostashiro) [inline account_invoice_delivered_amount/models/account_invoice_line.py]:
```suggestion
                line.delivered_amount = line.price_total
```

File: account_invoice_delivered_amount/models/account_invoice_line.py [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+                continue
+            qty_delivered = sum(line.sale_line_ids.mapped("qty_delivered"))
+            if qty_delivered > 0:
+                line.is_delivered = True
+                line.delivered_amount = qty_delivered * line.price_total
```
Comment (yostashiro) [inline account_invoice_delivered_amount/models/account_invoice_line.py]:
```suggestion
                currency = line.invoice_id and line.invoice_id.currency_id or None
                price = line.price_unit * (1 - (line.discount or 0.0) / 100.0)
                taxes = False
                if line.invoice_line_tax_ids:
                    taxes = line.invoice_line_tax_ids.compute_all(price, currency, line.quantity_delivered, product=line.product_id, partner=line.invoice_id.partner_id)
                delivered_subtotal = taxes["total_excluded"] if taxes else line.quantity_delivered * price
                line.delivered_amount = taxes["total_included"] if taxes else delivered_subtotal
```
To replicate the logic of the standard method https://github.com/odoo/odoo/blob/cd580dd60545766acd0a02ea9dc89e2c9c5de743/addons/account/models/account_invoice.py#L1431-L1437.  A bit more complicated than one would imagine...

**Review by yostashiro — CHANGES_REQUESTED** [review]

**Review by yostashiro — APPROVED** [review]

**Review by yostashiro — APPROVED** [review]

**Pull request merged**